File: ddcalc.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <time.h>

void ddcalc(float* data, int dataCount, int dataDim, float* distances) {
    for(int from = 0; from < dataCount; ++from) {
        for(int to = from + 1; to < dataCount; to++) {
            float distance = 0;
            for (int i = 0; i < dataDim; ++i) {
                distance += pow(data[from * dataDim + i] - data[to * dataDim + i], 2);
            }
            distance = sqrt(distance);
            distances[from * dataDim + to] = distance;
        }
    }
}

int main(int argc, char const *argv[])
{
    clock_t start, end;

    int elements = 20000;
    int dimension = 5;



    start = clock();
    srand(time(NULL));
    float* data = (float*) malloc(sizeof(float) * elements * dimension);
    float* distances = (float*) malloc(sizeof(float) * elements * elements);
    for(int i = 0; i < elements * dimension; ++i) data[i] = rand() / (float) RAND_MAX;
    end = clock();
    printf("data build time: %f\n", (end - start) / (double) CLOCKS_PER_SEC);

    start = clock();
    ddcalc(data, elements, dimension, distances);
    end = clock();
    printf("ddcalc time: %f\n", (end - start) / (double) CLOCKS_PER_SEC);

    return 0;
}

File: ddcalc.py
import numpy as np
import pyopencl as cl
import time


context = cl.create_some_context()

start = time.time()
queue = cl.CommandQueue(context)
program = cl.Program(
    context,
    '''
    __kernel void ddcalc(
        __global const float* data,
        const int dataCount,
        const int dataDim,
        __global float* distances
    ) {
        int from = get_global_id(0);
        int to = get_global_id(1);

        float distance = 0;
        for (int i = 0; i < dataDim; ++i) {
            distance += pow(data[from * dataDim + i] - data[to * dataDim + i], 2);
        }
        distances[from * dataCount + to] = sqrt(distance);
    }
    '''
).build()
end = time.time()
print(f'queue and kernel time: {end - start}')

elements = 20_000
dimension = 5

start = time.time()
data = np.random.rand(elements, dimension).astype(np.float32)
data_gpu = cl.Buffer(context, cl.mem_flags.READ_ONLY | cl.mem_flags.COPY_HOST_PTR, hostbuf=data)
distances = np.empty(elements ** 2, np.float32).reshape((elements, elements))
distances_gpu = cl.Buffer(context, cl.mem_flags.WRITE_ONLY, distances.nbytes)
end = time.time()
print(f'data build time: {end - start}')

start = time.time()
program.ddcalc(queue, (elements, elements), None, data_gpu, np.int32(elements), np.int32(dimension), distances_gpu)
end = time.time()
print(f'ddcalc time: {end - start}')

start = time.time()
cl.enqueue_copy(queue, distances, distances_gpu)
end = time.time()
print(f'copy back time: {end - start}')
